Update dashboard JSON and fix dashboard example

petervdonovan opened the pull request:
~~This PR makes the following changes:~~
- ~~Update the dashboard to include fixes for objectively wrong data that used to appear in the reaction execution timelines. These fixes come from the mixed-criticality dashboard.~~
- ~~Rework the time series panel which is now called "Maximum lag by logical time." Notably, this changes it so that the x-axis is logical time, not physical time. When I was working on the mixed-criticality examples, I found that this make the panel more interpretable. This improvement comes from the mixed-criticality dashboard.~~
- ~~Remove the panel called "Logical time by reactor," which we have discussed on Zulip and which we seem to have agreed is uninformative and difficult to understand. This improvement also comes from the mixed-criticality dashboard.~~
- ~~Update the dashboard with descriptions to clarify the basic meaning of certain visualizations. This is a response to a comment from this morning that one of our users wanted documentation for the dashboard.~~
- ~~Change some variable names in the example programs to avoid naming collisions. A more robust and general fix that uses IDs to distinguish between different programs that use the same FQNs is left for a future PR.~~
- ~~Update the C example so that deadline violations sometimes occur. This is done simply to sanity-check that the deadline violation metrics do work.~~

**EDIT: see CHANGELOG.md for the updated changelist.**

I tested these changes locally, and they seem to work on my machine. I used LFC commit `2948f1ff8b6a95bdce678ce515c6eea3fa5a4999`.

The JSON of this version of the dashboard should work with the SDV demo. It does not contain uids that will be invalid. I believe that it can safely be copied over to that repo.

petervdonovan committed:
Initial patch

petervdonovan committed:
Update example publishers

petervdonovan committed:
Add descriptions

This is a response to comments that it is not clear exactly what the
figures in the dashboard signify.

lhstrh commented on InfluxPublisher/README.md:
````diff
@@ -6,15 +6,15 @@ This folder contains Lingua Franca reactors that publish to InfluxDB via a socke
+   ```
 
-    Open Grafana on `http://0.0.0.0:3000` and open the dashboard "Telegraf Test Publishers".
+   Open Grafana on `http://0.0.0.0:3000` and open the dashboard "Telegraf Test Publishers".
````
Iwai-san reported:
> Open Grafana on http://0.0.0.0:3000 and open the dashboard "Telegraf Test Publishers".
-> http://localhost:3000

petervdonovan replied:
OK, done

petervdonovan committed:
Address bug report

elgeeko1 commented on InfluxPublisher/README.md:
````diff
@@ -6,15 +6,15 @@ This folder contains Lingua Franca reactors that publish to InfluxDB via a socke
+   ```
 
-    Open Grafana on `http://0.0.0.0:3000` and open the dashboard "Telegraf Test Publishers".
+   Open Grafana on `http://localhost:3000` and open the dashboard "Telegraf Test Publishers".
````
Curious why this change?

petervdonovan replied:
See https://github.com/xronos-inc/xronos-dashboard/pull/3#discussion_r1698997833. That's all I know.

elgeeko1 commented on InfluxPublisher/README.md:
````diff
@@ -6,15 +6,15 @@ This folder contains Lingua Franca reactors that publish to InfluxDB via a socke
-    cd py && lfc src/InfluxPublisher.lf
-    ```
+   ```bash
+   cd c && lfc src/InfluxPublisher.lf; cd ..
````
Ah, good cathch. I prefer `(cd c && lfc src/InfluxPublisher.lf)` since it's easier to combine with other commands, but just food for thought.

petervdonovan replied:
Sounds good, README.md has been updated with the improved Bash commands.

elgeeko1 commented on InfluxPublisher/c/src/InfluxPublisher.lf:
```diff
@@ -117,7 +123,7 @@ reactor SineWave(sample_rate_ms = 100 ms) {
-    publisher = new InfluxLinePublisher()
-    producer = new SineWave()
-    producer.measurement -> publisher.line
+    c_publisher = new InfluxLinePublisher()
```
Curious why this change? It makes sense from a naming standpoint, but wanted to know if there was another reason.

petervdonovan replied:
This is the workaround that we discussed on Wednesday morning. It is this change (described above):
> Change some variable names in the example programs to avoid naming collisions. A more robust and general fix that uses IDs to distinguish between different programs that use the same FQNs is left for a future PR.

elgeeko1 reviewed (changes requested):
One change to request here. The test publishers are designed to be templates that can be used for user-level publishing to InfluxDB. I think by adding deadlines we're mixing templates and examples, where the deadlines make sense only for examples, and don't make sense for a user who wants to have a template for how to publish to InfluxDB.

Instead, could we create a separate `examples` folder that has a publisher with deadline misses?

elgeeko1 commented:
I'm looking at the updated tracing dashboard now.

1. The tooltips are *very* helpful, as are the more detailed names for the panels.
2. The graph "Maximum lag time by logical time", I'm not sure I understand what is the meaning of lag in logical time with respect to the logical clock? I'm used to thinking of lag between logical time and physical time. Can you help me understand why this metric matters? (Also: the logarithmic scale is very clean)
3. I notice the update rate is now 1s, which we support with our latest `grafana.ini` configuration, but I'm surprised it works in the tracing dashboard. Can the Influx queries keep up with this? What did you learn when moving from a 5s update to a 1s update?
4. As I'm looking at the WCET diagram, I'm thinking of a new feature which is to show execution time as a percentage of total execution time of the application during the window, a proxy for CPU utilization. This is a feature request and not something we should build now.
5. Reaction executions, if I read this correctly, is showing a really valuable view. The behavior planner for example 
<img width="1330" alt="image" src="https://github.com/user-attachments/assets/617bf4d8-632b-4556-b222-91fbcd32b2a6"> shows some reactions executing nonstop. Is this what you would expect?
6. See the graph below. I see only green, what does this mean?
<img width="1303" alt="image" src="https://github.com/user-attachments/assets/128b9013-cac8-4b7b-b06b-9858faa6e675">
7. I see no deadline misses for the AVP demo, is this what you would expect?

I think it would be helpful to review the AVP demo together with you.

petervdonovan commented:
Thanks for your comments. I'll look at items 2 and 3 tomorrow morning. For item 2, I think I need to clarify the "description" and edit the title so that it is clear that the lag is physical time minus logical time, as usual. For item 3, probably you are right, the refresh rate should be reverted back to 5s; the only difference is how pleasant it is to look at.

Item 4 I will keep in mind for later.

For item 6, I'll take another look at the "description" for that panel. It should explain that higher opacity corresponds to a later logical time.

Items 5 and 7 I am not sure about. I agree that it might be good to look at it together.

petervdonovan committed:
Address comments 2, 3, and 6

petervdonovan committed:
Improve bash commands in README

petervdonovan committed:
Separate deadlines example from starter code

petervdonovan commented:
> Instead, could we create a separate examples folder that has a publisher with deadline misses?

Yep, that is the better way. Done in [2944bdb](https://github.com/xronos-inc/xronos-dashboard/pull/3/commits/2944bdb7a55d203032242dd7e2586f27e7e3c81e).

petervdonovan committed:
Fix bad artifacts at start and end of time range

petervdonovan committed:
Update time range to 15s instead of 10

petervdonovan committed:
Fix minor bug in initial value

petervdonovan committed:
Take microstep into account in "opacity indicates logical time" timeline

petervdonovan committed:
Start a changelog.

petervdonovan committed:
Add missing notes to CHANGELOG.md

petervdonovan committed:
Add screenshots demonstrating the bugfix

petervdonovan committed:
Add more screenshots showing the bugfix

petervdonovan committed:
Fix another very subtle bug

Apparently it is possible for two invocations of `lf_time_physical` to
return the same time when it is invoked from different threads? So this
means that it is merely increasing, and not strictly increasing?

petervdonovan commented:
@elgeeko1 @cmnrd I think this PR is once again ready for review. Thanks for your patience as it turned out I did not do due diligence the first time around. I think that the bugs that were apparent in `Multi_Tread.lf` and `Single_Tread.lf` have been ironed out.

petervdonovan commented on CHANGELOG.md:
```diff
@@ -0,0 +1,30 @@
+  - This also changes the panel titled "Reaction executions (opacity indicates logical time and microstep)" to use opacity to indicate logical time and microstep instead of using color to indicate logical time. This can make the panel difficult to read without zooming in, but for technical reasons, it is necessary in order to enable this bugfix.
+  - Example 1:
+    - Before this bugfix, reaction executions are often dropped or reported as having the wrong execution times, seemingly at random:
+      ![after](documentation/images/changelog/2024-08-05/multi_tread_master.png)
```
```suggestion
      ![before](documentation/images/changelog/2024-08-05/multi_tread_master.png)
```

elgeeko1 committed:
Rename folder 'documentation' to 'docs'

elgeeko1 committed:
Rename folder 'examples/Deadlines' to 'examples/deadline-misses'

elgeeko1 committed:
Rename example to DeadlineMisses.lf and adjust timing.

petervdonovan committed:
Make panel "Reaction WCET" more readable

petervdonovan committed:
Update CHANGELOG.md to reflect "Reaction WCET" fix

petervdonovan committed:
Simplify deadline-misses example

This replaces a busy-wait with a sleep. I am not sure if it fixes
anything, but I am trying it in response to a 75ms lag that I cannot
reproduce. In any case it seems like a simplification that could make
this less error-prone.

petervdonovan committed:
Add back accidentally deleted comment

petervdonovan committed:
Tiny update

cmnrd commented on CHANGELOG.md:
```diff
@@ -0,0 +1,35 @@
+      ![before](docs/images/changelog/2024-08-05/single_tread_master.png)
+    - After:
+      ![after](docs/images/changelog/2024-08-05/single_tread_corrected.png)
+- Fix a bug involving unclear and non-unique names that made the "Reaction WCET" panel difficult to interpret. This panel now reports distinct data for each reaction and does not aggregate data across distinct reactions in the same process.
```
```suggestion
- Fix a bug involving unclear and non-unique names that made the "Reaction WCET" panel difficult to interpret. This panel now reports distinct data for each reaction and does not aggregate data across distinct reactions in the same federate.
```

cmnrd reviewed (approved):
I run this again and can confirm that this indeed fixes the bugs that we observed.

Should we move the InfluxPublisher example also to examples and perhaps give it a more descriptive name (e.g. SineWaveGenerator)?

petervdonovan committed:
Update CHANGELOG.md

Co-authored-by: Christian Menard <Christian.Menard@gmx.de>

rcakella commented:
@elgeeko1, I tried this today and it looks good to me.
